core/dnsserver/server.go

```diff
@@ -59,6 +62,15 @@ func NewServer(addr string, group []*Config) (*Server, error) {
 		var stack middleware.Handler
 		for i := len(site.Middleware) - 1; i >= 0; i-- {
 			stack = site.Middleware[i](stack)
+			if s.trace == nil && stack.Name() == "trace" {
+				// we have to stash away the middleware, not the
+				// Tracer object, because the Tracer won't be initialized yet
+				if t, ok := stack.(trace.Trace); ok {
+					s.trace = t
+				} else {
+					log.Println("[WARNING] Invalid type found for trace middleware. This is a bug.")
```

Comment:
This happens when we add a middleware with `Name()` is trace, but isn't a `trace.Trace`? Think it is save to drop the else here. The users of CoreDNS won't know what to do anyway.

(We could add a tests for middleware having unique names)